src/ppx_irmin/README.md

```diff
@@ -0,0 +1,99 @@
+# ppx_irmin
+
+PPX extension for automatically generating Irmin generics.
+
+[![Build Status](https://travis-ci.com/CraigFe/ppx_irmin.svg?branch=master)](https://travis-ci.com/CraigFe/ppx_irmin)
+## Overview
+
+`ppx_irmin` automatically generates Irmin generics (values of type `_ Irmin.Type.t`)
+corresponding to type declarations in your code. For example:
+
+```ocaml
+type tree =
+  | Branch of tree * bool option * tree
+  | Leaf of int32 * string [@@deriving irmin]
+```
+
+will be expanded to:
+
+```ocaml
+type tree = (* as above *)
+
+let tree_t =
+  let open Irmin.Type in
+  mu (fun tree_t ->
+      variant "tree" (fun branch leaf -> function
+          | Branch (x1, x2, x3) -> branch (x1, x2, x3)
+          | Leaf   (x1, x2)     -> leaf (x1, x2))
+      |~ case1 "Branch" (triple tree_t (option bool) tree_t) (fun (x1, x2, x3) -> Branch (x1, x2, x3))
+      |~ case1 "Leaf"   (pair int32 string)                  (fun (x1, x2) -> Leaf (x1, x2))
+      |> sealv)
+```
+
+## Installation and usage
+
+You can install `ppx_irmin` using [opam](https://opam.ocaml.org/):
+
+```
+$ git clone git@github.com:CraigFe/ppx_irmin
+$ opam pin --yes ./ppx_irmin
+```
+
+If you're using the [dune](https://github.com/ocaml/dune) build system, add the following field to
+your `library`, `executable` or `test` stanza:
+
+```
+(preprocess (pps ppx_irmin))
+```
+
+You can now use `[@@deriving irmin]` after a type declaration in your code to automatically derive
+an Irmin generic with the same name.
+
+
+## Specifics
+
+`ppx_irmin` supports all of the type combinators exposed in the
+[Irmin.Type](https://docs.mirage.io/irmin/Irmin/Type/index.html) module (basic types, records,
+variants (plain and closed polymorphic), recursive types etc.). Irmin types must fully grounded (no
```

Comment:
I guess *must*  is a current limitation of the implementation, not a general limitation. Maybe be a bit more explicit about it (e.g. *not currently supported* or something like this).

Reply:
Indeed, the [analogous tool in Haskell](http://dreixel.net/research/pdf/gdmh.pdf) does support higher-kinded generics via a `Par` combinator. Clarified in https://github.com/mirage/irmin/pull/919/commits/e77dffbd68874ca09485577755ec8e5de5fc05e7; let me know what you think.